AssertionError raised in the case of a function without inputs.

I'm trying to make a pipeline with a loading function that takes no inputs. Something like the following:

```python
def func_a() -> float:
    return 10

def func_b(a: float) -> float:
    return a + 30

Pipeline(
    [
        PipeFunc(func_a, "a"),
        PipeFunc(func_b, "b")
    ]
)
```

When I execute this, I get the following assertion error:
```python
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[13], line 1
----> 1 Pipeline(
      2     [
      3         PipeFunc(func_a, "a"),
      4         PipeFunc(func_b, "b")
      5     ]
      6 )

File ~/.local/opt/micromamba/envs/pipeline/lib/python3.11/site-packages/pipefunc/_pipeline.py:129, in Pipeline.__init__(self, functions, lazy, debug, profile, cache_type, cache_kwargs, scope, default_resources)
    127     else:
    128         mapspec = None
--> 129     self.add(f, mapspec=mapspec)
    130 self._cache_type = cache_type
    131 self._cache_kwargs = cache_kwargs

File ~/.local/opt/micromamba/envs/pipeline/lib/python3.11/site-packages/pipefunc/_pipeline.py:212, in Pipeline.add(self, f, mapspec)
    209     f.debug = self.debug
    211 self._clear_internal_cache()  # reset cache
--> 212 self._validate()
    213 return f

File ~/.local/opt/micromamba/envs/pipeline/lib/python3.11/site-packages/pipefunc/_pipeline.py:942, in Pipeline._validate(self)
    940 _validate_scopes(self.functions)
    941 _check_consistent_defaults(self.functions, output_to_func=self.output_to_func)
--> 942 self._validate_mapspec()

File ~/.local/opt/micromamba/envs/pipeline/lib/python3.11/site-packages/pipefunc/_pipeline.py:954, in Pipeline._validate_mapspec(self)
    952         raise ValueError(msg)
    953 validate_consistent_axes(self.mapspecs(ordered=False))
--> 954 self._autogen_mapspec_axes()

File ~/.local/opt/micromamba/envs/pipeline/lib/python3.11/site-packages/pipefunc/_pipeline.py:997, in Pipeline._autogen_mapspec_axes(self)
    995 def _autogen_mapspec_axes(self) -> set[PipeFunc]:
    996     """Generate `MapSpec`s for functions that return arrays with ``internal_shapes``."""
--> 997     root_args = self.topological_generations.root_args
    998     mapspecs = self.mapspecs(ordered=False)
    999     non_root_inputs = _find_non_root_axes(mapspecs, root_args)

File ~/.local/opt/micromamba/envs/pipeline/lib/python3.11/functools.py:1001, in cached_property.__get__(self, instance, owner)
    999 val = cache.get(self.attrname, _NOT_FOUND)
   1000 if val is _NOT_FOUND:
-> 1001     val = self.func(instance)
   1002     try:
   1003         cache[self.attrname] = val

File ~/.local/opt/micromamba/envs/pipeline/lib/python3.11/site-packages/pipefunc/_pipeline.py:981, in Pipeline.topological_generations(self)
    978 if not generations:
    979     return Generations([], [])
--> 981 assert all(isinstance(x, str | _Bound | _Resources) for x in generations[0])
    982 assert all(isinstance(x, PipeFunc) for gen in generations[1:] for x in gen)
    983 root_args = [x for x in generations[0] if isinstance(x, str)]

AssertionError: 
```

I can remove the assertion error by inserting a dummy variable in `func_a` but this feels like a buggy assertion.

--- a/pipefunc/_pipeline.py
+++ b/pipefunc/_pipeline.py
@@ -974,44 +974,35 @@ class Pipeline:
         contains the root arguments, while the subsequent generations contain
         the functions in topological order.
         """
-        generations = list(nx.topological_generations(self.graph))
+        parameterless_functions = [f for f in self.functions if not f.parameters]
+        graph = self.graph
+        if parameterless_functions:
+            graph = self.graph.copy()
+            for i, f in enumerate(parameterless_functions):
+                graph.add_edge(i, f)
+        else:
+            graph = self.graph
+
+        generations = list(nx.topological_generations(graph))
         if not generations:
             return Generations([], [])
 
-        root_args = []
-        parameterless_funcs = []
-        function_lists = []
+        root_args: list[str] = []
+        function_lists: list[list[PipeFunc]] = []
         for i, generation in enumerate(generations):
             function_lists.append([])
             for x in generation:
-                if isinstance(x, str):
-                    assert i == 0
+                if i == 0 and isinstance(x, str):
                     root_args.append(x)
-                elif i == 0:
-                    if isinstance(x, PipeFunc):
-                        parameterless_funcs.append([x])
-                    elif not isinstance(x, _Bound | _Resources):
-                        msg = f"Unexpected type: {type(x)}"
-                        raise TypeError(msg)
+                elif i == 0 and isinstance(x, _Bound | _Resources | int):
+                    pass
                 else:
+                    assert isinstance(x, PipeFunc)
                     function_lists[-1].append(x)
 
         # Remove empty lists
         function_lists = [lst for lst in function_lists if lst]
         assert all(isinstance(x, PipeFunc) for gen in function_lists for x in gen)
-
-        def insert_at_index(f: PipeFunc, function_lists: list[list[PipeFunc]]) -> int | None:
-            for i, gen in enumerate(function_lists):
-                if any(set(f.output_name) & set(func.parameters) for func in gen):
-                    return max(i - 1, 1)
-            return None
-
-        for f in parameterless_funcs:
-            index = insert_at_index(f, function_lists)
-            if index is not None:
-                function_lists[index].append(f)
-            else:
-                function_lists.insert(0, [f])
 
         return Generations(root_args, function_lists)
 

--- a/pipefunc/_pipeline.py
+++ b/pipefunc/_pipeline.py
@@ -982,8 +982,7 @@ class Pipeline:
         parameterless_funcs = []
         function_lists = []
         for i, generation in enumerate(generations):
-            if i == 0 or not function_lists[-1]:
-                function_lists.append([])
+            function_lists.append([])
             for x in generation:
                 if isinstance(x, str):
                     assert i == 0
@@ -997,6 +996,8 @@ class Pipeline:
                 else:
                     function_lists[-1].append(x)
 
+        # Remove empty lists
+        function_lists = [lst for lst in function_lists if lst]
         assert all(isinstance(x, PipeFunc) for gen in function_lists for x in gen)
 
         def insert_at_index(f: PipeFunc, function_lists: list[list[PipeFunc]]) -> int | None:
@@ -1010,7 +1011,8 @@ class Pipeline:
             if index is not None:
                 function_lists[index].append(f)
             else:
-                function_lists.append([f])
+                function_lists.insert(0, [f])
+
         return Generations(root_args, function_lists)
 
     @functools.cached_property

--- a/pipefunc/_pipeline.py
+++ b/pipefunc/_pipeline.py
@@ -994,7 +994,7 @@ class Pipeline:
         root_args: list[str] = []
         function_lists: list[list[PipeFunc]] = []
         for i, generation in enumerate(generations):
-            function_lists.append([])
+            generation_functions: list[PipeFunc] = []
             for x in generation:
                 if i == 0 and isinstance(x, str):
                     root_args.append(x)
@@ -1003,10 +1003,10 @@ class Pipeline:
                     pass
                 else:
                     assert isinstance(x, PipeFunc)
-                    function_lists[-1].append(x)
+                    generation_functions.append(x)
+            if generation_functions:
+                function_lists.append(generation_functions)
 
-        # Remove any empty generations
-        function_lists = [lst for lst in function_lists if lst]
         assert all(isinstance(x, PipeFunc) for gen in function_lists for x in gen)
 
         return Generations(root_args, function_lists)

--- a/pipefunc/_pipeline.py
+++ b/pipefunc/_pipeline.py
@@ -1007,8 +1007,6 @@ class Pipeline:
             if generation_functions:
                 function_lists.append(generation_functions)
 
-        assert all(isinstance(x, PipeFunc) for gen in function_lists for x in gen)
-
         return Generations(root_args, function_lists)
 
     @functools.cached_property

--- a/pipefunc/_pipeline.py
+++ b/pipefunc/_pipeline.py
@@ -969,16 +969,20 @@ class Pipeline:
     def topological_generations(self) -> Generations:
         """Return the functions in the pipeline grouped by topological generation.
 
-        Simply calls `networkx.topological_generations` on the `pipeline.graph`. Then
-        groups the functions in the pipeline by generation. The first generation
-        contains the root arguments, while the subsequent generations contain
-        the functions in topological order.
+        This method uses `networkx.topological_generations` on the pipeline graph to group
+        functions by their dependency order. The result includes:
+        - Root arguments: Initial inputs to the pipeline.
+        - Function generations: Subsequent groups of functions in topological order.
+
+        Nullary functions (those without parameters) are handled specially to ensure
+        they're included in the generations rather than treated as root arguments.
         """
-        parameterless_functions = [f for f in self.functions if not f.parameters]
-        graph = self.graph
-        if parameterless_functions:
+        nullary_functions = [f for f in self.functions if not f.parameters]
+        if nullary_functions:
+            # Handle nullary functions by adding placeholder edges.
+            # This ensures they're included in the generations rather than as root arguments.
             graph = self.graph.copy()
-            for i, f in enumerate(parameterless_functions):
+            for i, f in enumerate(nullary_functions):
                 graph.add_edge(i, f)
         else:
             graph = self.graph
@@ -995,12 +999,13 @@ class Pipeline:
                 if i == 0 and isinstance(x, str):
                     root_args.append(x)
                 elif i == 0 and isinstance(x, _Bound | _Resources | int):
+                    # Skip special first-generation nodes that aren't root arguments
                     pass
                 else:
                     assert isinstance(x, PipeFunc)
                     function_lists[-1].append(x)
 
-        # Remove empty lists
+        # Remove any empty generations
         function_lists = [lst for lst in function_lists if lst]
         assert all(isinstance(x, PipeFunc) for gen in function_lists for x in gen)
 

--- a/pipefunc/_pipeline.py
+++ b/pipefunc/_pipeline.py
@@ -978,10 +978,40 @@ class Pipeline:
         if not generations:
             return Generations([], [])
 
-        assert all(isinstance(x, str | _Bound | _Resources) for x in generations[0])
-        assert all(isinstance(x, PipeFunc) for gen in generations[1:] for x in gen)
-        root_args = [x for x in generations[0] if isinstance(x, str)]
-        return Generations(root_args, generations[1:])
+        root_args = []
+        parameterless_funcs = []
+        function_lists = []
+        for i, generation in enumerate(generations):
+            if i == 0 or not function_lists[-1]:
+                function_lists.append([])
+            for x in generation:
+                if isinstance(x, str):
+                    assert i == 0
+                    root_args.append(x)
+                elif i == 0:
+                    if isinstance(x, PipeFunc):
+                        parameterless_funcs.append([x])
+                    elif not isinstance(x, _Bound | _Resources):
+                        msg = f"Unexpected type: {type(x)}"
+                        raise TypeError(msg)
+                else:
+                    function_lists[-1].append(x)
+
+        assert all(isinstance(x, PipeFunc) for gen in function_lists for x in gen)
+
+        def insert_at_index(f: PipeFunc, function_lists: list[list[PipeFunc]]) -> int | None:
+            for i, gen in enumerate(function_lists):
+                if any(set(f.output_name) & set(func.parameters) for func in gen):
+                    return max(i - 1, 1)
+            return None
+
+        for f in parameterless_funcs:
+            index = insert_at_index(f, function_lists)
+            if index is not None:
+                function_lists[index].append(f)
+            else:
+                function_lists.append([f])
+        return Generations(root_args, function_lists)
 
     @functools.cached_property
     def sorted_functions(self) -> list[PipeFunc]:

--- a/tests/test_pipefunc.py
+++ b/tests/test_pipefunc.py
@@ -1940,3 +1940,33 @@ def test_calling_add_with_autogen_mapspec():
         internal_shapes={"foo_out": (3,)},
     )
     assert results["bar_out"].output.tolist() == [1, 4, 9]
+
+
+def test_parameterless_pipefunc() -> None:
+    @pipefunc(output_name="c")
+    def f():
+        return 1
+
+    assert f() == 1
+
+    pipeline = Pipeline([f])
+    assert pipeline() == 1
+    assert pipeline.topological_generations.root_args == []
+    assert pipeline.topological_generations.function_lists == [[pipeline["c"]]]
+
+    @pipefunc(output_name="d")
+    def g():
+        return 2
+
+    @pipefunc(output_name="e")
+    def h(c, d):
+        return c + d
+
+    pipeline = Pipeline([f, g, h])
+
+    assert pipeline() == 3
+    assert pipeline.topological_generations.root_args == []
+    assert pipeline.topological_generations.function_lists == [
+        [pipeline["c"], pipeline["d"]],
+        [pipeline["e"]],
+    ]

--- a/tests/test_pipefunc.py
+++ b/tests/test_pipefunc.py
@@ -1953,6 +1953,8 @@ def test_parameterless_pipefunc() -> None:
     assert pipeline() == 1
     assert pipeline.topological_generations.root_args == []
     assert pipeline.topological_generations.function_lists == [[pipeline["c"]]]
+    r = pipeline.map({})
+    assert r["c"].output == 1
 
     @pipefunc(output_name="d")
     def g():
@@ -1970,3 +1972,5 @@ def test_parameterless_pipefunc() -> None:
         [pipeline["c"], pipeline["d"]],
         [pipeline["e"]],
     ]
+    r = pipeline.map({})
+    assert r["e"].output == 3

